<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Population Science - 4&nbsp; Sequence Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./network.html" rel="next">
<link href="./geodemographics.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Sequence Analysis</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Population Science</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/fcorowe/r4ps" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Download" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-download"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./Population-Science.pdf">
            <i class="bi bi-bi-file-pdf pe-1"></i>
          Download PDF
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="./Population-Science.epub">
            <i class="bi bi-bi-journal pe-1"></i>
          Download ePub
          </a>
        </li>
    </ul>
    <a href="" title="Share" id="sidebar-tool-dropdown-1" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-1">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro-population-science.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introducing Population Science</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./geodemographics.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Geodemographics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sequence-analysis.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Sequence Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./network.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Network Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sentiment-analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Sentiment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic-modelling.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Topic Modelling</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./longitudinal-1.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Modelling Time</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./longitudinal-2.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Assessing Interventions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./machine-learning.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Machine Learning</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-sets.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Data sets</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#dependencies" id="toc-dependencies" class="nav-link active" data-scroll-target="#dependencies"><span class="toc-section-number">4.1</span>  Dependencies</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="toc-section-number">4.2</span>  Data</a>
  <ul class="collapse">
  <li><a href="#data-wrangling" id="toc-data-wrangling" class="nav-link" data-scroll-target="#data-wrangling"><span class="toc-section-number">4.2.1</span>  Data wrangling</a></li>
  <li><a href="#exploratory-data-analysis" id="toc-exploratory-data-analysis" class="nav-link" data-scroll-target="#exploratory-data-analysis"><span class="toc-section-number">4.2.2</span>  Exploratory data analysis</a></li>
  </ul></li>
  <li><a href="#application" id="toc-application" class="nav-link" data-scroll-target="#application"><span class="toc-section-number">4.3</span>  Application</a>
  <ul class="collapse">
  <li><a href="#defining-outcome-process" id="toc-defining-outcome-process" class="nav-link" data-scroll-target="#defining-outcome-process"><span class="toc-section-number">4.3.1</span>  Defining outcome process</a></li>
  <li><a href="#optimal-matching" id="toc-optimal-matching" class="nav-link" data-scroll-target="#optimal-matching"><span class="toc-section-number">4.3.2</span>  Optimal matching</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"><span class="toc-section-number">4.3.3</span>  Clustering</a></li>
  <li><a href="#visualising" id="toc-visualising" class="nav-link" data-scroll-target="#visualising"><span class="toc-section-number">4.3.4</span>  Visualising</a></li>
  </ul></li>
  <li><a href="#questions" id="toc-questions" class="nav-link" data-scroll-target="#questions"><span class="toc-section-number">4.4</span>  Questions</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/fcorowe/r4ps/edit/main/sequence-analysis.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-chp4" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Sequence Analysis</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>This chapter illustrates the use of sequence analysis and <a href="https://www.worldpop.org">WorldPop</a> raster data to identify trajectories of population decline in Ukraine. Developed in Biology for the analysis of DNA sequencing, sequence analysis offers a novel approach to generate a more holistic understanding of population decline trajectories capturing differences in the ordering, frequency, timing and occurrence of population decline. The <em>longitudinal</em> and <em>categorical</em> nature is a key feature of the data that can be analysed using sequence analysis. In this chapter, We first show how to manipulate gridded, raster population to create a spatial data frame, and explore national and sub-national population trends and spatial structure of population change. We describe the process of implementing sequence analysis to identify trajectories of population decline in a four-stage process. We first define different levels of population change. Second, we apply measure the difference or similarity between individual area-level trajectories. Third, we use an unsupervised machine learning clustering algorithm to identify representative trajectories. Fourth, we use different visualisation tools to understand key features encoded in trajectories.</p>
<p>The chapter is based on the following references:</p>
<p><span class="citation" data-cites="gabadinho2011">Gabadinho et al. (<a href="references.html#ref-gabadinho2011" role="doc-biblioref">2011</a>)</span> describe the functionalities of the <code>TraMineR</code> package to visualise and analysis categorical sequence data;</p>
<p><span class="citation" data-cites="newsham2022a">Newsham and Rowe (<a href="references.html#ref-newsham2022a" role="doc-biblioref">2022b</a>)</span>, <span class="citation" data-cites="gonzález-leonardo2023">González-Leonardo, Newsham, and Rowe (<a href="references.html#ref-gonzález-leonardo2023" role="doc-biblioref">2023</a>)</span> provide empirical applications to identify and study trajectories of population decline across Europe and in Spain;</p>
<p><span class="citation" data-cites="tatem2017">Tatem (<a href="references.html#ref-tatem2017" role="doc-biblioref">2017</a>)</span> provide an overview of the <em>WorldPop</em> data project;</p>
<p><span class="citation" data-cites="patias2021">Patias, Rowe, and Arribas-Bel (<a href="references.html#ref-patias2021" role="doc-biblioref">2021</a>)</span>, <span class="citation" data-cites="patias2021a">Patias et al. (<a href="references.html#ref-patias2021a" role="doc-biblioref">2021</a>)</span> provide good examples of the use of sequence analysis to define trajectories of inequality and urban development.</p>
<section id="dependencies" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="dependencies"><span class="header-section-number">4.1</span> Dependencies</h2>
<p>We use the libraries below. Note that to use the <code>theme_tufte2()</code> used for <code>ggplot()</code> objects in this chapter, you need to call the file <code>data-viz-themes.R</code> in the repository.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># data manipulation</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># spatial data manipulation</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stars)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># download world pop data</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(wpgpDownloadR) <span class="co"># you may need to install this package running `install.packages("devtools")` `devtools::install_github("wpgp/wpgpDownloadR")'</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># data visualisation</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggseqplot) <span class="co"># may need to install by running `devtools::install_github("maraab23/ggseqplot")`</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># sequence analysis</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TraMineR)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster analysis</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Key packages to this chapter are <code>TraMineR</code>,<code>stars</code> and <code>ggseqplot</code>. <code>TraMineR</code> is the go-to package in social sciences for exploring, analysing and rendering sequences based on categorical data. <code>stars</code> is designed to handle spatio-temporal data in the form of dense arrays, with space and time as dimensions. <code>stars</code> provides classes and methods for reading, manipulating, plotting and writing data cubes. It is a very powerful package. It interacts nicely with <code>sf</code> and is suggested to be superior to <code>raster</code> and <code>terra</code>, which are also known for their capacity to work with multilayer rasters. <code>stars</code> is suggested to <a href="https://r-spatial.github.io/stars/">deal with more complex data types</a> and <a href="https://www.r-bloggers.com/2021/05/a-comparison-of-terra-and-stars-packages/">be faster</a> than <code>raster</code> and <code>terra</code>. <code>ggseqplot</code> provides functionality to visualise categorical sequence data based on <code>ggplot</code> capabilities. This differs from <code>TraMineR</code> which is based on the base function <code>plot</code>. We prefer <code>ggseqplot</code> for the wide usage of <code>ggplot</code> as a data visualisation tool in R.</p>
</section>
<section id="data" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="data"><span class="header-section-number">4.2</span> Data</h2>
<p>The key aim of this chapter is to define representative trajectories of population decline using sequence analysis and WorldPop data. We use WorldPop data for the period extending from 2000 to 2020. WorldPop offers open access gridded population estimates at a high spatial resolution for all countries in the world. WoldPop produces these gridded datasets using a top-down (i.e.&nbsp;dissagregating administrative area counts into smaller grid cells) or bottom-up (i.e.&nbsp;interpolating data from counts from sample locations into grid cells) approach. You can learn about about these approaches and the data available from <a href="https://www.worldpop.org">WorldPop</a>.</p>
<p>WorldPop population data are available in various formats:</p>
<ul>
<li><p>Two spatial resolutions: 100m and 1km;</p></li>
<li><p>Constrained and unconstrained counts to built settlement areas;</p></li>
<li><p>Adjusted or unadjusted to United Nations’ (UN) national population counts;</p></li>
<li><p>Two formats i.e.&nbsp;<code>tiff</code> and <code>csv</code> formats.</p></li>
</ul>
<p>We use annual 1km gridded, UN adjusted, unconstrained population count data for Ukraine during 2000-2021 in tiff format. We use tiff formats to illustrate the manipulation of raster data. Such skills will come handy if you ever decide to work with satellite imagery or image data in general.</p>
<p>Before calling the data, let’s see how we can use <code>wpgpDownloadR</code> package. Let’s browse the data catalogue.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wpgpListCountries</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in readLines(con, n = 1): incomplete final line found on
'/var/folders/pt/t0w6_05j22dctdkh3pgkpfjc0000gp/T//RtmpNYKRFn/wpgpDatasets.md5'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  ISO ISO3            Country
1 643  RUS             Russia
2 360  IDN          Indonesia
3 840  USA      United States
4 850  VIR Virgin_Islands_U_S
5 304  GRL          Greenland
6 156  CHN              China</code></pre>
</div>
</div>
<p>By using the ISO3 country code, let’s look for the available datasets for Ukraine.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wpgpListCountryDatasets</span>(<span class="at">ISO3 =</span> <span class="st">"UKR"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in readLines(con, n = 1): incomplete final line found on
'/var/folders/pt/t0w6_05j22dctdkh3pgkpfjc0000gp/T//RtmpNYKRFn/wpgpDatasets.md5'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>     ISO ISO3 Country Covariate
232  804  UKR Ukraine  ppp_2000
481  804  UKR Ukraine  ppp_2001
730  804  UKR Ukraine  ppp_2002
979  804  UKR Ukraine  ppp_2003
1228 804  UKR Ukraine  ppp_2004
1477 804  UKR Ukraine  ppp_2005
                                                                                                                                                                 Description
232  Estimated total number of people per grid-cell 2000 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)
481  Estimated total number of people per grid-cell 2001 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)
730  Estimated total number of people per grid-cell 2002 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)
979  Estimated total number of people per grid-cell 2003 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)
1228 Estimated total number of people per grid-cell 2004 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)
1477 Estimated total number of people per grid-cell 2005 The dataset is available to download in Geotiff format at a resolution of 3 arc (approximately 100m at the equator)</code></pre>
</div>
</div>
<p>The <code>wpgpDownloadR</code> package includes 100m resolution data. To keep things efficient, we use 1km gridded population counts from the <a href="https://hub.worldpop.org/geodata/listing?id=75">WorldPop data page</a>. Obtain population data for Ukraine 2000-2020. We start by reading the set of tiff files using the <code>read_stars</code> function from the <code>star</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a list of file names</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> fs<span class="sc">::</span><span class="fu">dir_ls</span>(<span class="st">"./data/sequence-analysis/raster"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>file_list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>./data/sequence-analysis/raster/ukr_ppp_2000_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2001_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2002_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2003_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2004_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2005_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2006_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2007_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2008_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2009_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2010_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2011_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2012_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2013_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2014_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2015_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2016_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2017_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2018_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2019_1km_Aggregated_UNadj.tif
./data/sequence-analysis/raster/ukr_ppp_2020_1km_Aggregated_UNadj.tif</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read a list of raster data</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>pop_raster <span class="ot">&lt;-</span> <span class="fu">read_stars</span>(file_list, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We map the data for 2000 to get a quick understanding of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pop_raster[<span class="dv">1</span>], <span class="at">col =</span> <span class="fu">inferno</span>(<span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>downsample set to 2</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next we read shapefile of administrative boundaries in the form of polygons. We obtain these data from the <a href="https://gadm.org">GADM website</a>. GADM provides maps and spatial data for individuals countries at the national and sub-national administrative divisions. In this chapter, we will work with these data as they come directly from the website which provides a more realistic and similar context to which you will probably come across in the “real-world”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read spatial data frame</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>ukr_shp <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">"./data/sequence-analysis/ukr_shp/gadm41_UKR_2.shp"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">st_simplify</span>(., <span class="co"># simplify boundaries for efficiency</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">preserveTopology =</span> T,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">dTolerance =</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span>  <span class="co"># 1km</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  sf<span class="sc">::</span><span class="fu">st_make_valid</span>(.) <span class="sc">%&gt;%</span> </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fortify</span>(.) <span class="sc">%&gt;%</span>  <span class="co"># turns maps into a data frame so they can more easily be plotted with ggplot2</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">st_transform</span>(., <span class="st">"EPSG:4326"</span>) <span class="co"># set projection system</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `gadm41_UKR_2' from data source 
  `/Users/PIETROST/Library/CloudStorage/Dropbox/GitHub/r4ps/data/sequence-analysis/ukr_shp/gadm41_UKR_2.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 629 features and 13 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: 22.14045 ymin: 44.38597 xmax: 40.21807 ymax: 52.37503
Geodetic CRS:  WGS 84</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ukr_shp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Simple feature collection with 629 features and 13 fields (with 1 geometry empty)
Geometry type: GEOMETRY
Dimension:     XY
Bounding box:  xmin: 22.14519 ymin: 44.38681 xmax: 40.21807 ymax: 52.375
Geodetic CRS:  WGS 84
First 10 features:
       GID_2 GID_0 COUNTRY   GID_1   NAME_1 NL_NAME_1            NAME_2
1          ?   UKR Ukraine       ?        ?         ?                 ?
2  UKR.1.1_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська        Cherkas'ka
3  UKR.1.2_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська       Cherkas'kyi
4  UKR.1.3_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська   Chornobaivs'kyi
5  UKR.1.4_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська     Chyhyryns'kyi
6  UKR.1.5_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська       Drabivs'kyi
7  UKR.1.6_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська Horodyshchens'kyi
8  UKR.1.7_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська       Kamians'kyi
9  UKR.1.8_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська         Kanivs'ka
10 UKR.1.9_1   UKR Ukraine UKR.1_1 Cherkasy Черкаська        Kanivs'kyi
         VARNAME_2 NL_NAME_2      TYPE_2                     ENGTYPE_2 CC_2
1                ?        NA           ?                            NA   NA
2               NA        NA Mis'ka Rada City of Regional Significance   NA
3               NA        NA       Raion                      District   NA
4  Chornobayivskyi        NA       Raion                      District   NA
5               NA        NA       Raion                      District   NA
6               NA        NA       Raion                      District   NA
7  Gorodyschenskyi        NA       Raion                      District   NA
8               NA        NA       Raion                      District   NA
9               NA        NA       Misto                          City   NA
10              NA        NA       Raion                      District   NA
     HASC_2                       geometry
1         ? POLYGON ((30.59574 50.40547...
2  UA.CK.CM POLYGON ((32.1715 49.43881,...
3  UA.CK.CR POLYGON ((32.03393 49.49881...
4  UA.CK.CB POLYGON ((32.17991 49.44486...
5  UA.CK.CY POLYGON ((32.26144 49.20893...
6  UA.CK.DR POLYGON ((32.41852 49.83724...
7  UA.CK.HO POLYGON ((31.56959 49.42509...
8  UA.CK.KN POLYGON ((32.19797 49.20946...
9  UA.CK.KM MULTIPOLYGON (((31.4459 49....
10 UA.CK.KR POLYGON ((31.5851 49.62482,...</code></pre>
</div>
</div>
<p>Let’s have a quick look at the resolution of the administrative areas we will be working. The areas below represent areas at the administrative area level 2 in the spatial data frame <code>ukr_shp</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ukr_shp<span class="sc">$</span>geometry)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We ensure that the <code>pop_raster</code> object is in the same projection system as <code>ukr_shp</code>. So we can make both objects to work together.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pop_raster <span class="ot">&lt;-</span> <span class="fu">st_transform</span>(pop_raster, <span class="fu">st_crs</span>(ukr_shp))                      </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="data-wrangling" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="data-wrangling"><span class="header-section-number">4.2.1</span> Data wrangling</h3>
<p>For our application, we want to work with administrative areas for three reasons. First, public policy and planning decisions are often made based on administrative areas. These are the areas local governments have jurisdiction, represent and can exert power. Second, migration is a key component of population change and hence directly determines population decline. At a small area, residential mobility may also impact patterns of population potentially adding more complexity and variability to the process. Third, WorldPop data are modelled population estimates with potentially high levels of uncertainty or errors in certain locations. Our aim is to mitigate the potential impacts of these errors.</p>
<p>We therefore recommend working with aggregated data. We aggregate the 1km gridded population data to administrative areas in Ukraine. We use <code>system.time</code> to time the duration of the proccess of aggregation which could take some time depending on your local computational environment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system.time</span>({</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>popbyarea_df <span class="ot">=</span> <span class="fu">aggregate</span>(<span class="at">x =</span> pop_raster, </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">by =</span> ukr_shp, </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">FUN =</span> sum, </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">na.rm =</span> <span class="cn">TRUE</span>) </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   user  system elapsed 
 66.391  10.794  77.545 </code></pre>
</div>
</div>
<p><strong>Sub-national population</strong></p>
<p>The chunk code above returns a list of raster data. We want to create a spatial data frame containing population counts for individual sub-national areas and years. We achieve this by running the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a function to bind the population data frame to the shapefile</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>add_population <span class="ot">&lt;-</span> <span class="cf">function</span>(x) <span class="fu">mutate</span>(ukr_shp, </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">population =</span> x)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain sub-national population counts</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>ukr_eshp <span class="ot">&lt;-</span> <span class="fu">lapply</span>(popbyarea_df, add_population)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co"># create a dataframe with sub-national populations</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>select_pop <span class="ot">&lt;-</span> <span class="cf">function</span>(x) dplyr<span class="sc">::</span><span class="fu">select</span>(x, GID_2, NAME_2, population)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>population_df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(ukr_eshp, select_pop) <span class="sc">%&gt;%</span> </span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(rbind, .)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>population_df<span class="sc">$</span>year <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">2000</span>, <span class="dv">2020</span>), <span class="at">times =</span> <span class="dv">1</span>, <span class="at">each =</span> <span class="fu">nrow</span>(ukr_shp))</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(population_df) <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(population_df), <span class="at">by=</span><span class="dv">1</span>), <span class="at">times =</span> <span class="dv">1</span>)</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co"># sub-national spatial data frame</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>population_df </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Simple feature collection with 13209 features and 4 fields (with 21 geometries empty)
Geometry type: GEOMETRY
Dimension:     XY
Bounding box:  xmin: 22.14519 ymin: 44.38681 xmax: 40.21807 ymax: 52.375
Geodetic CRS:  WGS 84
First 10 features:
       GID_2            NAME_2 population                       geometry year
1          ?                 ?  301849.00 POLYGON ((30.59574 50.40547... 2000
2  UKR.1.1_1        Cherkas'ka  280917.39 POLYGON ((32.1715 49.43881,... 2000
3  UKR.1.2_1       Cherkas'kyi   89116.78 POLYGON ((32.03393 49.49881... 2000
4  UKR.1.3_1   Chornobaivs'kyi   50096.24 POLYGON ((32.17991 49.44486... 2000
5  UKR.1.4_1     Chyhyryns'kyi   36646.73 POLYGON ((32.26144 49.20893... 2000
6  UKR.1.5_1       Drabivs'kyi   42467.86 POLYGON ((32.41852 49.83724... 2000
7  UKR.1.6_1 Horodyshchens'kyi   49886.59 POLYGON ((31.56959 49.42509... 2000
8  UKR.1.7_1       Kamians'kyi   35587.28 POLYGON ((32.19797 49.20946... 2000
9  UKR.1.8_1         Kanivs'ka   14406.93 MULTIPOLYGON (((31.4459 49.... 2000
10 UKR.1.9_1        Kanivs'kyi   37495.04 POLYGON ((31.5851 49.62482,... 2000</code></pre>
</div>
</div>
<p><strong>National population</strong></p>
<p>We also create a data frame providing population counts at the national level.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain national population counts</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>population_count <span class="ot">&lt;-</span> <span class="fu">map_dbl</span>(ukr_eshp, <span class="sc">~</span>.x <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>          <span class="fu">pull</span>(population) <span class="sc">%&gt;%</span> </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>          <span class="fu">sum</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">%&gt;%</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># change labels</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(population_count) <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"population"</span>)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(population_count) <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">2000</span>, <span class="dv">2020</span>, <span class="at">by=</span><span class="dv">1</span>), <span class="at">times =</span> <span class="dv">1</span>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>population_count<span class="sc">$</span>year <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">2000</span>, <span class="dv">2020</span>, <span class="at">by=</span><span class="dv">1</span>), <span class="at">times =</span> <span class="dv">1</span>)</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># national annual population counts</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>population_count</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     population year
2000   47955683 2000
2001   47520197 2001
2002   47094225 2002
2003   46700872 2003
2004   46330322 2004
2005   46011048 2005
2006   45734099 2006
2007   45502336 2007
2008   45286748 2008
2009   45090608 2009
2010   44923112 2010
2011   44744969 2011
2012   44593427 2012
2013   44424702 2013
2014   44250993 2014
2015   44068072 2015
2016   43856852 2016
2017   43622605 2017
2018   43391259 2018
2019   43140679 2019
2020   42880388 2020</code></pre>
</div>
</div>
</section>
<section id="exploratory-data-analysis" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="exploratory-data-analysis"><span class="header-section-number">4.2.2</span> Exploratory data analysis</h3>
<p>Now we are ready to start analysing the data. Before building complexity on our analysis, conducting some exploratory data analysis to understand the data is generally a good starting point, particularly given the multi-layer nature of the data at hand - capturing space, time and population levels.</p>
<p><strong>National patterns</strong></p>
<p>We first analyse national population trends. We want to know to what extent the population of Ukraine has declined over time over the last 20 years. An effective way to do this is to compute summary statistics and visualise the data. Below we look at year-to-year changes in population levels and as a percentage change. By using <code>patchwork</code>, we combine two plots into a single figure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualise national population trends</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>pop_level_p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(population_count, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> population<span class="sc">/</span><span class="dv">1000000</span> )) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tufte2</span>() <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">48</span>) <span class="sc">+</span> </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Population </span><span class="sc">\n</span><span class="st">(million)"</span>,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Year"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualise percentage change in population</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>pop_percent_p <span class="ot">&lt;-</span> population_count <span class="sc">%&gt;%</span> </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">pct_change =</span> ( ( population <span class="sc">-</span> <span class="dv">47955683</span>) <span class="sc">/</span> <span class="dv">47955683</span>) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> pct_change )) <span class="sc">+</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tufte2</span>() <span class="sc">+</span> </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Population </span><span class="sc">\n</span><span class="st">percentage change (%)"</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Year"</span>) </span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>pop_level_p <span class="sc">|</span> pop_percent_p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><strong>Sub-national</strong></p>
<p>Population losses are likely vary across the country. From previous research we know that rural and less well connected areas tend to lose population through the internal migration of young individuals as they move for work and job opportunities <span class="citation" data-cites="rowe2016">(<a href="references.html#ref-rowe2016" role="doc-biblioref">Rowe, Corcoran, and Bell 2016</a>)</span>. We also know that they tend to move to large, densely populated cities where these opportunities are concentrated and because they also offer a wide variety of amenities and activities. Cities tend to work as accelarators enabling fast career development and occupational progression <span class="citation" data-cites="fielding1992">(<a href="references.html#ref-fielding1992" role="doc-biblioref">Fielding 1992</a>)</span>. Though, we have also seen the shrinkage of populations in cities, particularly in eastern European countries <span class="citation" data-cites="turok2007">(<a href="references.html#ref-turok2007" role="doc-biblioref">Turok and Mykhnenko 2007</a>)</span>.</p>
<p>To examine the patterns of sun-national population losses, we compute two summary measures: (1) annual percentage change in population; and, (2) overall percentage change in population between 2000 and 2021. We start by looking the overall percentage change as it is easier to visualise. To this end, we categorise our measure of overall percentage change into seven different classes. Based on previous work by <span class="citation" data-cites="gonzález-leonardo2023">González-Leonardo, Newsham, and Rowe (<a href="references.html#ref-gonzález-leonardo2023" role="doc-biblioref">2023</a>)</span>, we classify changes into high decline (<span class="math inline">\(\leq\)</span> -3), decline (&gt; -3 and <span class="math inline">\(\leq\)</span> -1.5), moderate decline (&gt; -1.5 and <span class="math inline">\(\leq\)</span> -0.3), stable (&lt; -0.3 and &lt; 0.3), moderate growth (<span class="math inline">\(\geq\)</span> 0.3 and &lt; 1.5), growth (<span class="math inline">\(\geq\)</span> 1.5 and &lt; 3) and high growth (<span class="math inline">\(\geq\)</span> 3). Let’s first create the measures of population change.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute population change metrics</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>population_df <span class="ot">&lt;-</span> population_df <span class="sc">%&gt;%</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(GID_2) <span class="sc">%&gt;%</span> </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="sc">-</span>year, <span class="at">.by_group =</span> <span class="cn">TRUE</span> ) <span class="sc">%&gt;%</span> </span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pct_change =</span> ( population <span class="sc">/</span> <span class="fu">lead</span>(population) <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="co"># rate of population change</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">pct_change_2000_21 =</span> ( population[year <span class="sc">==</span> <span class="st">"2020"</span>] <span class="sc">/</span> population[year <span class="sc">==</span> <span class="st">"2000"</span>] <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">*</span> <span class="dv">100</span>, <span class="co"># overall rate of change</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ave_pct_change_2000_21 =</span> <span class="fu">mean</span>(pct_change, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> </span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s map the overall percentage change in population between 2000 and 2020. We see a wide spread pattern of population decline across Ukraine. We observe a large spatial cluster of high population decline across the country with moderate population decline in some areas. Administrative areas containing large cities seem to record overall population growth between 2000 and 2020, potentially absorbing population movements from the rest of the country. What else do you think may be driving population growth in cities? And in contrast, what do you think is contributing to population decline in most of Ukraine?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set colours</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#7f3b08"</span>, <span class="st">"#b35806"</span>, <span class="st">"#e08214"</span>, <span class="st">"#faf0e6"</span>, <span class="st">"#8073ac"</span>, <span class="st">"#542788"</span>, <span class="st">"#2d004b"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co"># reverse order</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rev</span>(cols)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>population_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>( year <span class="sc">==</span> <span class="dv">2020</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>(pct_change_2000_21) <span class="sc">%&gt;%</span> </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">ove_pop_class =</span> <span class="fu">case_when</span>( pct_change_2000_21 <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">3</span> <span class="sc">~</span> <span class="st">'high_decline'</span>,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&lt;=</span> <span class="sc">-</span><span class="fl">1.5</span> <span class="sc">&amp;</span> pct_change_2000_21 <span class="sc">&gt;</span> <span class="sc">-</span><span class="dv">3</span> <span class="sc">~</span> <span class="st">'decline'</span>,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&lt;=</span> <span class="sc">-</span>.<span class="dv">3</span> <span class="sc">&amp;</span> pct_change_2000_21 <span class="sc">&gt;</span> <span class="sc">-</span><span class="fl">1.5</span> <span class="sc">~</span> <span class="st">'moderate_decline'</span>,</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&gt;</span> <span class="sc">-</span><span class="fl">0.3</span> <span class="sc">&amp;</span> pct_change_2000_21 <span class="sc">&lt;</span> <span class="fl">0.3</span> <span class="sc">~</span> <span class="st">'stable'</span>,</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&gt;=</span> <span class="fl">0.3</span> <span class="sc">&amp;</span> pct_change_2000_21 <span class="sc">&lt;</span> <span class="fl">1.5</span> <span class="sc">~</span> <span class="st">'moderate_growth'</span>,</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&gt;=</span> <span class="fl">1.5</span> <span class="sc">&amp;</span> pct_change_2000_21 <span class="sc">&lt;</span> <span class="dv">3</span> <span class="sc">~</span> <span class="st">'growth'</span>,</span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>                           pct_change_2000_21 <span class="sc">&gt;=</span> <span class="dv">3</span> <span class="sc">~</span> <span class="st">'high_growth'</span>),</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">ove_pop_class =</span> <span class="fu">factor</span>(ove_pop_class, </span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"high_decline"</span>, <span class="st">"decline"</span>, <span class="st">"moderate_decline"</span>, <span class="st">"stable"</span>, <span class="st">"moderate_growth"</span>, <span class="st">"growth"</span>, <span class="st">"high_growth"</span>) )</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> ove_pop_class)) <span class="sc">+</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">col =</span> <span class="st">"white"</span>, <span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cols,</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>                    <span class="at">name =</span> <span class="st">"Population change"</span>) <span class="sc">+</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_map_tufte</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now that we have understanding of population changes over the whole 2000-2020 period. Let’s try to understand how different places arrive to different outcomes. A way to do this is to look at the evolution of population changes. Different trajectories of population change could underpin the outcomes of population change that we observe today. Current outcomes could be the result of a consistent pattern of population decline over the last 20 years. They could be the result of acceleration in population loss after a major natural or war event, or they could reflect a gradual process of erosion. We visualise way to get an understanding of this is to analyse annual percentage population changes across individual areas. We use a Hovmöller Plot as illustrated by <span class="citation" data-cites="rowe2022e">Rowe and Arribas-Bel (<a href="references.html#ref-rowe2022e" role="doc-biblioref">2022</a>)</span> for the analysis of spatio-temporal data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>population_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>( ave_pct_change_2000_21 <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tail</span>(., <span class="dv">40</span><span class="sc">*</span><span class="dv">21</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> ., </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span> year, <span class="at">y=</span> <span class="fu">reorder</span>(NAME_2, pct_change), <span class="at">fill=</span> pct_change)) <span class="sc">+</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis</span>(<span class="at">name=</span><span class="st">"Population"</span>, <span class="at">option =</span><span class="st">"plasma"</span>, <span class="at">begin =</span> .<span class="dv">2</span>, <span class="at">end =</span> .<span class="dv">8</span>, <span class="at">direction =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tufte2</span>() <span class="sc">+</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span> <span class="fu">paste</span>(<span class="st">" "</span>), <span class="at">x=</span><span class="st">"Year"</span>, <span class="at">y=</span><span class="st">"Area"</span>) <span class="sc">+</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>)) <span class="sc">+</span> </span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The Hovmöller Plot shows that most of the selected areas tend to experience annual population decline, with varying spells of population growth. Percentage population changes range between 1 and -2.5. We also observe areas with consistent trajectories of annual population decline, like Zolochivs’kyi and Barvinkivs’kyi, and areas with strong decline in the first few years between 2000 and 2005 but moderate decline later on, such as Novovorontsovk’kyi. Yet, Hovmöller Plots provide a limited understanding of the annual population changes for a handful of areas at the time and it is therefore difficult to identify systematic representative patterns. Here we have selected 40 areas of a total of 629. Displaying the total number of areas in a Hovmöller Plot will not produce readable results. Even if that was the case, it would be difficult to identify systematic patterns. As we will seek to persuade you below, sequence analysis provides a very novel way to define representative trajectories in the data, identify systematic patterns and extract distinctive features characterising those trajectories.</p>
</section>
</section>
<section id="application" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="application"><span class="header-section-number">4.3</span> Application</h2>
<p>Next, we focus on the application of sequence analysis to identify representative trajectories of population decline at the sub-national level between 2000 and 2020 in Ukraine. Intuitively, sequence analysis can be seen as a four-stage process. First, it requires the definition of longitudinal categorical outcome. Second, it measures the dissimilarity of individual sequences via a process known as optimal matching (OM). Third, it uses these dissimilarity measures to define a typology of representative trajectories using unsupervised machine learning clustering techniques. Fourth, trajectories can be visualised and their distinctive features can be measured. Below we describe the implementation of each stage to identify representative trajectories of population decline.</p>
<section id="defining-outcome-process" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="defining-outcome-process"><span class="header-section-number">4.3.1</span> Defining outcome process</h3>
<p>Sequence analysis requires longitudinal categorical data as an input. We therefore classify our population count data into distinct categorical categories, henceforth referred to as <em>states</em> of population change. We compute the annual percentage rate of population change for individual areas and use these rates to measure the extent and pace of population change. The annual rate of population change is computed as follows:</p>
<p><span class="math display">\[
{p(t1) - p(t0) \over p(t0)}*100
\]</span></p>
<p>where: <span class="math inline">\(p(t0)\)</span> is the population at year t0 and <span class="math inline">\(p(t1)\)</span> is the population at t + 1.</p>
<p>As previously, we differentiate areas of high decline, decline, moderate decline, stable, moderate growth, growth and high growth. For the analysis, we focus on areas recording population losses between 2000 and 2020. The histogram shows the magnitude and distribution of population decline over this period. We observe that most occurrences of decline are moderate around zero, while very few exceed 5%.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select areas reporting losses between 2000 and 2020</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>population_loss_df <span class="ot">&lt;-</span> population_df <span class="sc">%&gt;%</span> </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>( pct_change_2000_21 <span class="sc">&lt;</span> <span class="dv">0</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot distribution of percentage change </span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>population_loss_df <span class="sc">%&gt;%</span> </span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(pct_change  <span class="sc">&lt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span>  ) <span class="sc">+</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha=</span><span class="fl">0.8</span>, <span class="at">colour=</span><span class="st">"black"</span>, <span class="at">fill=</span><span class="st">"lightblue"</span>, <span class="fu">aes</span>(<span class="at">x =</span> pct_change)) <span class="sc">+</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tufte2</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Next we classify the annual percentage of population change into our seven states.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove 2000 as it has no observations of population change</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>population_loss_df <span class="ot">&lt;-</span> population_loss_df <span class="sc">%&gt;%</span> </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>( year <span class="sc">!=</span> <span class="dv">2000</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co"># clasify data</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>population_loss_df <span class="ot">&lt;-</span> population_loss_df <span class="sc">%&gt;%</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pop_class =</span> <span class="fu">case_when</span>( pct_change <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">3</span> <span class="sc">~</span> <span class="st">'high_decline'</span>,</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&lt;=</span> <span class="sc">-</span><span class="fl">1.5</span> <span class="sc">&amp;</span> pct_change <span class="sc">&gt;</span> <span class="sc">-</span><span class="dv">3</span> <span class="sc">~</span> <span class="st">'decline'</span>,</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&lt;=</span> <span class="sc">-</span>.<span class="dv">3</span> <span class="sc">&amp;</span> pct_change <span class="sc">&gt;</span> <span class="sc">-</span><span class="fl">1.5</span> <span class="sc">~</span> <span class="st">'moderate_decline'</span>,</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&gt;</span> <span class="sc">-</span><span class="fl">0.3</span> <span class="sc">&amp;</span> pct_change <span class="sc">&lt;</span> <span class="fl">0.3</span> <span class="sc">~</span> <span class="st">'stable'</span>,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&gt;=</span> <span class="fl">0.3</span> <span class="sc">&amp;</span> pct_change <span class="sc">&lt;</span> <span class="fl">1.5</span> <span class="sc">~</span> <span class="st">'moderate_growth'</span>,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&gt;=</span> <span class="fl">1.5</span> <span class="sc">&amp;</span> pct_change <span class="sc">&lt;</span> <span class="dv">3</span> <span class="sc">~</span> <span class="st">'growth'</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>                           pct_change <span class="sc">&gt;=</span> <span class="dv">3</span> <span class="sc">~</span> <span class="st">'high_growth'</span>)</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="optimal-matching" class="level3" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="optimal-matching"><span class="header-section-number">4.3.2</span> Optimal matching</h3>
<p>We measure the extent of dissimilarity between individual sequence of population decline. To this end, we used a sequence analysis technique, OM, which computes distances between sequences as a function of the number of transformations required to make sequences identical. Two sets of operations are generally used: (1) insertion/deletion (known as indel) and (2) substitution operations. Both of these operations represent the cost of transforming one sequence into another. These costs are challenging to define and below we discuss what is generally used in empirical work. Intuitively, the idea of OM is to estimate the cost of transforming one sequence into another so that the greater the cost to make two sequences identical, the greater the dissimilarity and <em>vice versa</em>.</p>
<p><em>Indel operations</em> involve the addition or removal of an element within the sequence and substitution operations are the replacement of one element for another. Each of these operations is assigned a cost, and the distance between two sequences is defined as the minimum cost to transform one sequence to another <span class="citation" data-cites="abbott2000sequence">(<a href="references.html#ref-abbott2000sequence" role="doc-biblioref">Abbott and Tsay 2000</a>)</span>. By default, indel costs are set to 1. To illustrate indel operations, let’s consider an example of sequences of annual population change for three areas during 2000 and 2003. The sequences are identical, except for 2003. In this case, indel operations involve the cost of transforming the status <em>stable</em> in the sequence for area 1 to <em>high decline</em> in the sequence for area 2, and thus this operation would return a cost is 2. Why 2? It is 2 because you would need to delete <em>stable</em> and add <em>high decline</em>. Now, let’s try the cost of transforming the status <em>stable</em> in the sequence for area 1 to the status in the sequence for area 3 using indel operations. What is the cost? The answer is 1 because we only need to delete <em>stable</em> to make it identical.</p>
<table class="table">
<thead>
<tr class="header">
<th>Area</th>
<th>2000</th>
<th>2001</th>
<th>2002</th>
<th>2003</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>decline</td>
<td>decline</td>
<td>decline</td>
<td>stable</td>
</tr>
<tr class="even">
<td>2</td>
<td>decline</td>
<td>decline</td>
<td>decline</td>
<td>high decline</td>
</tr>
<tr class="odd">
<td>3</td>
<td>decline</td>
<td>decline</td>
<td>decline</td>
<td>-</td>
</tr>
</tbody>
</table>
<p><em>Substitution operations</em> or costs represent transition costs; that is, the cost for substituting each state with another. Substitution costs are defined in one of two ways <span class="citation" data-cites="salmela-aro2011">(<a href="references.html#ref-salmela-aro2011" role="doc-biblioref">Salmela-Aro et al. 2011</a>)</span>. One approach is the theory-driven approach. In such approach, substitution costs are grounded in theory suggesting that, for example, transforming state 1 to state 2 should have a greater cost than transforming state 1 to state 3, or performing the opposite operation i.e.&nbsp;transforming state 2 to state 1. An example could be that it is more financially costly to transition from <em>full-time employment</em> to <em>full-time education</em> than transition from <em>full-time education</em> to <em>full-time employment</em>.</p>
<p>A second approach and most commonly used in empirical work is a data-driven approach. In this approach, substitution costs are empirically derived from transition rates between states. The cost of substitution is inversely related to the frequency of observed transitions within the data. This means that infrequent transitions between states have a higher substitution cost. For example, as we will see, transitions from the state of high decline to high growth are rarer than from high growth to high decline in Ukraine. The transition rate between state <span class="math inline">\(i\)</span> and state <span class="math inline">\(j\)</span> is the probability of observing state <span class="math inline">\(j\)</span> at time <span class="math inline">\(t1\)</span> given that the state <span class="math inline">\(i\)</span> is observed at time <span class="math inline">\(t\)</span> for <span class="math inline">\(i \neq j\)</span>. The substitution cost between states <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> is computed as:</p>
<p><span class="math display">\[
2 - {p(i | j) - p(j | i)}
\]</span> where <span class="math inline">\(p(i | j)\)</span> is the transition rate between state <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>.</p>
<p>To implement OM, we first need to rearrange the structure of our data from long to wide format. You can now see now that individual rows represent areas (column 1) and columns from 2 to 21 represent years.</p>
<ul>
<li>see <span class="citation" data-cites="rowe2022e">Rowe and Arribas-Bel (<a href="references.html#ref-rowe2022e" role="doc-biblioref">2022</a>)</span> for a description on different spatio-temporal data structures and their manipulation using tidyverse principles.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transform from long to wide format</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>wide_population_loss_df <span class="ot">&lt;-</span> population_loss_df <span class="sc">%&gt;%</span> </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(GID_2) <span class="sc">%&gt;%</span> </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(year, <span class="at">.by_group =</span> <span class="cn">TRUE</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">id_cols =</span>  GID_2,</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_from =</span> <span class="st">"year"</span>,</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_from =</span> <span class="st">"pop_class"</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>wide_population_loss_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 571 × 21
   GID_2   `2001` `2002` `2003` `2004` `2005` `2006` `2007` `2008` `2009` `2010`
   &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt; 
 1 UKR.1.… moder… moder… moder… moder… moder… moder… stable moder… moder… stable
 2 UKR.1.… decli… moder… decli… decli… decli… moder… moder… moder… moder… moder…
 3 UKR.1.… moder… moder… moder… moder… moder… moder… moder… moder… moder… moder…
 4 UKR.1.… moder… moder… moder… moder… moder… moder… moder… moder… moder… moder…
 5 UKR.1.… decli… decli… moder… decli… moder… moder… moder… moder… moder… moder…
 6 UKR.1.… decli… decli… moder… decli… moder… moder… moder… moder… moder… moder…
 7 UKR.1.… moder… moder… moder… moder… moder… moder… moder… moder… moder… moder…
 8 UKR.1.… moder… moder… moder… moder… moder… moder… moder… moder… moder… moder…
 9 UKR.1.… moder… moder… moder… moder… moder… moder… moder… moder… moder… moder…
10 UKR.1.… decli… decli… decli… decli… decli… decli… decli… decli… moder… decli…
# … with 561 more rows, and 10 more variables: `2011` &lt;chr&gt;, `2012` &lt;chr&gt;,
#   `2013` &lt;chr&gt;, `2014` &lt;chr&gt;, `2015` &lt;chr&gt;, `2016` &lt;chr&gt;, `2017` &lt;chr&gt;,
#   `2018` &lt;chr&gt;, `2019` &lt;chr&gt;, `2020` &lt;chr&gt;</code></pre>
</div>
</div>
<p>Once the data frame has been reshaped into a wide format, we define the data as a state sequence object using the R package <code>TraMineR</code>. Key here is to appropriately define the labels and an appropriate palette of colours. Depending on the patterns you are seeking to capture a diverging, sequential or qualitative colour palette may be more appropriate. For this chapter, we use a diverging colour palette as we want to effectively represent areas experiencing diverging patterns of population decline or growth.</p>
<blockquote class="blockquote">
<p>Note: various types of sequence data representation exist in <code>TraMineR</code>. These representations vary in the way they capture states or events. Chapter 4 in <span class="citation" data-cites="gabadinho2009mining">Gabadinho et al. (<a href="references.html#ref-gabadinho2009mining" role="doc-biblioref">2009</a>)</span> describes the various representations that <code>TraMineR</code> can handle. In any case, the state sequence representation used in this chapter is the most commonly used and internal format used by <code>TraMineR</code>. Hence we focus on it.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># alphabet</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>seq.alphab <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"high_growth"</span>, <span class="st">"growth"</span>, <span class="st">"moderate_growth"</span>, <span class="st">"stable"</span>, <span class="st">"moderate_decline"</span>, <span class="st">"decline"</span>, <span class="st">"high_decline"</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># labels</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>seq.lab <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"High growth"</span>, <span class="st">"Growth"</span>, <span class="st">"Moderate growth"</span>, <span class="st">"Stable"</span>, <span class="st">"Moderate decline"</span>, <span class="st">"Decline"</span>, <span class="st">"High decline"</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># define state sequence object</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>seq.cl <span class="ot">&lt;-</span> <span class="fu">seqdef</span>(wide_population_loss_df, </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                 <span class="dv">2</span><span class="sc">:</span><span class="dv">21</span>, </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">alphabet =</span> seq.alphab,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">labels =</span> seq.lab,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>                 <span class="at">cnames =</span> <span class="fu">c</span>(<span class="st">"2001"</span>, <span class="st">"2002"</span>, <span class="st">"2003"</span>, <span class="st">"2004"</span>, <span class="st">"2005"</span>, <span class="st">"2006"</span>, <span class="st">"2007"</span>, <span class="st">"2008"</span>, <span class="st">"2009"</span>, <span class="st">"2010"</span>, <span class="st">"2011"</span>, <span class="st">"2012"</span>, <span class="st">"2013"</span>, <span class="st">"2014"</span>, <span class="st">"2015"</span>, <span class="st">"2016"</span>, <span class="st">"2017"</span>, <span class="st">"2018"</span>, <span class="st">"2019"</span>, <span class="st">"2020"</span>),</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">cpal =</span><span class="fu">c</span>(<span class="st">"1"</span> <span class="ot">=</span> <span class="st">"#7f3b08"</span>, </span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"#b35806"</span>,</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"#e08214"</span>,</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"#faf0e6"</span>,</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"#8073ac"</span>,</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"#542788"</span>,</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"#2d004b"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code> [&gt;] 7 distinct states appear in the data: </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     1 = decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     2 = growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     3 = high_decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     4 = high_growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     5 = moderate_decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     6 = moderate_growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     7 = stable</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> [&gt;] state coding:</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>       [alphabet]       [label]          [long label] </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     1  high_growth      high_growth      High growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     2  growth           growth           Growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     3  moderate_growth  moderate_growth  Moderate growth</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     4  stable           stable           Stable</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     5  moderate_decline moderate_decline Moderate decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     6  decline          decline          Decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>     7  high_decline     high_decline     High decline</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> [&gt;] 571 sequences in the data set</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code> [&gt;] min/max sequence length: 20/20</code></pre>
</div>
</div>
<p>Using the sequence data object, we create a state distribution plot to get an understanding of the data. The plot shows the distribution of areas across status of population change in individual years. The overall picture emerging from the plot is an overall pattern of population decline between 2000 and 2020, predominantly moderate decline and limited spells of high population decline or growth. This aligns with the predominant trajectory of population decline observed in Ukrain based on more aggregate data at the regional level <span class="citation" data-cites="newsham2022">(<a href="references.html#ref-newsham2022" role="doc-biblioref">Newsham and Rowe 2022a</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqplot</span>(seq.cl, </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">title=</span><span class="st">"State distribution plot"</span>, </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="st">"d"</span>,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">with.legend =</span> <span class="st">"right"</span>,</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">border =</span> <span class="cn">NA</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We now move on to compute the substitution costs for our population states. From the equation above, you may have realised that transition costs vary between 0 and 2, with the former indicating zero cost. The latter indicates the maximum cost of converting one state into another. The option <code>TRATE</code> in method states to derive costs from observed transition rates. That is the data-driven approach discussed above. From the matrix below, you can see that it is more costly to convert a status “high growth” to “moderate decline” than from “growth” to “moderate”. This makes sense. We expect gradual changes in population along a given trajectory if they were to occur due to natural causes.</p>
<blockquote class="blockquote">
<p>Note we are considering a fixed measure of transition rates. That means that we are using the whole dataset to compute an average transition rate between states. That assumes that the rate of change between states does not change over time. Yet, there may be good reasons to believe they do as areas move across different states. In empirical work, time varying transition rates are more often considered. That means we use temporal slices of the data to compute transition rates; for example, using data from 2001, 2002 and so on. In this way, we end up with potentially different transition rates for every year.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate transition rates</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>subs_costs <span class="ot">&lt;-</span> <span class="fu">seqsubm</span>(seq.cl, </span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">"TRATE"</span>,</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>                      <span class="co">#time.varying = TRUE</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>                      )</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>subs_costs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 high_growth   growth moderate_growth   stable moderate_decline
high_growth         0.000000 1.975000        1.940066 1.949203         1.949277
growth              1.975000 0.000000        1.867730 1.867946         1.634772
moderate_growth     1.940066 1.867730        0.000000 1.709425         1.547832
stable              1.949203 1.867946        1.709425 0.000000         1.501723
moderate_decline    1.949277 1.634772        1.547832 1.501723         0.000000
decline             1.847681 1.848052        1.826810 1.915740         1.643018
high_decline        1.288462 1.700000        1.773408 1.917223         1.805088
                  decline high_decline
high_growth      1.847681     1.288462
growth           1.848052     1.700000
moderate_growth  1.826810     1.773408
stable           1.915740     1.917223
moderate_decline 1.643018     1.805088
decline          0.000000     1.844570
high_decline     1.844570     0.000000</code></pre>
</div>
</div>
<p>To understand better the idea of substitution costs, we can have direct look at transition rates underpinning these costs. Transition rates can be computed via <code>seqtrate</code> . By definition, transition rates vary between 0 and 1, with zero indicating no probability of a transition occurring. One indicates a 100% probability of a transition taking place. Thus, for example, the matrix below tell us that there is a 5% probability of observing a transition from “high growth” to “moderate decline” in our sample. Examining transition rates could provide very valuable information about the process in analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>seq.trate <span class="ot">&lt;-</span> <span class="fu">seqtrate</span>(seq.cl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code> [&gt;] computing transition probabilities for states high_growth/growth/moderate_growth/stable/moderate_decline/decline/high_decline ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(seq.trate, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      [-&gt; high_growth] [-&gt; growth] [-&gt; moderate_growth]
[high_growth -&gt;]                  0.12        0.03                 0.05
[growth -&gt;]                       0.00        0.05                 0.11
[moderate_growth -&gt;]              0.01        0.02                 0.09
[stable -&gt;]                       0.00        0.00                 0.04
[moderate_decline -&gt;]             0.00        0.00                 0.02
[decline -&gt;]                      0.00        0.01                 0.02
[high_decline -&gt;]                 0.16        0.10                 0.18
                      [-&gt; stable] [-&gt; moderate_decline] [-&gt; decline]
[high_growth -&gt;]             0.05                  0.05         0.15
[growth -&gt;]                  0.13                  0.36         0.15
[moderate_growth -&gt;]         0.25                  0.43         0.15
[stable -&gt;]                  0.46                  0.42         0.05
[moderate_decline -&gt;]        0.08                  0.82         0.08
[decline -&gt;]                 0.03                  0.28         0.65
[high_decline -&gt;]            0.07                  0.19         0.15
                      [-&gt; high_decline]
[high_growth -&gt;]                   0.55
[growth -&gt;]                        0.20
[moderate_growth -&gt;]               0.05
[stable -&gt;]                        0.01
[moderate_decline -&gt;]              0.00
[decline -&gt;]                       0.01
[high_decline -&gt;]                  0.15</code></pre>
</div>
</div>
<p>Now we focus on the probably most important component of sequence analysis; that is, the calculation of dissimilarity. Recall our aim is to identify representative trajectories. To this end, we need a way to measure how similar or different sequences are - which is known as OM. Above, we described that we can use indel and substitution operations to measure the dissimilarity or costs between individual sequences. The code chunk implements OM based on indel and substitution operations. The algorithm takes an individual sequence and compares it with all of the sequences in the dataset, and identifies the sequence with the minimum cost i.e.&nbsp;the most similar sequence. The result of this computing intensive process is a distance matrix encoding the similarity or dissimilarity between individual sequences.</p>
<blockquote class="blockquote">
<p>For indel, <code>auto</code> sets the indel as <code>max(sm)/2</code> when sm is a matrix. For more details, run <code>?seqdist</code> on your console</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate a distance matrix</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>seq.om <span class="ot">&lt;-</span> <span class="fu">seqdist</span>(seq.cl,</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">method =</span> <span class="st">"OM"</span>, <span class="co"># specify the method</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">indel =</span> <span class="st">"auto"</span>, <span class="co"># specify indel costs</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">sm =</span> subs_costs) <span class="co"># specify substitution costs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As highlighted above, if you would like to apply varying substitution costs, you can do this directly here by using the option <code>method = DHD</code> .</p>
</section>
<section id="clustering" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="clustering"><span class="header-section-number">4.3.3</span> Clustering</h3>
<p>The resulting distance matrix from OM <code>seq.om</code> indicates the degree of similarity between individual sequences. To identify representative trajectories, we then need to a way to group together similar sequences to produce a typology, in this case of population decline trajectories. Unsupervised cluster analysis is generally used for this task. Trusting you have built an understanding of cluster analysis from the previous chapter, we will not provide an elaborate description here. If you would like to know more about cluster analysis, we recommend the introductory book by <span class="citation" data-cites="kaufman2009finding">Kaufman and Rousseeuw (<a href="references.html#ref-kaufman2009finding" role="doc-biblioref">2009</a>)</span>. We use a clustering method called <code>k-meloids</code> . This methods is known to be more robust to noise and outliers than the conventional k-means procedure <span class="citation" data-cites="backman2020">(<a href="references.html#ref-backman2020" role="doc-biblioref">Backman, Lopez, and Rowe 2020</a>)</span>. This is because the medoid algorithm clusters the data by minimising a sum of pair-wise dissimilarities <span class="citation" data-cites="kaufman2009finding">(<a href="references.html#ref-kaufman2009finding" role="doc-biblioref">Kaufman and Rousseeuw 2009</a>)</span>, rather than a sum of squared Euclidean distances. We run cluster analyses at different numbers of <code>k</code> starting from 2 to 20.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run PAMs</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  pam_sol <span class="ot">&lt;-</span> <span class="fu">pam</span>(seq.om, k)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then seek to determine the optimal number of clusters <code>k</code>. We use silhouette scores, but as we noted <a href="geodemographics.html"><span>Chapter&nbsp;3</span></a>, the optimal number of clusters is better determined by the user given the context and use case. It is an art. There is no wrong or right answer. As can be seen from the results below from the average silhouette score, two clusters is suggested as the optimal solution. However, we could argue that we gain very little from such coarse partition of the data. We suggest to take this as guidance and a starting point to look to identify an appropriate data partition. We suggest to visualise different solution and gain an understanding of what data get split and decide on whether the resulting patterns contribute to the understanding of the process at hand.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute average silhouette scores for all 20 cluster solutions</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>asw <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="dv">20</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  asw[k] <span class="ot">&lt;-</span> <span class="fu">pam</span>(seq.om, k) <span class="sc">$</span> silinfo <span class="sc">$</span> avg.width</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  k.best <span class="ot">&lt;-</span> <span class="fu">which.max</span>(asw)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"silhouette-optimal number of clusters:"</span>, k.best, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>silhouette-optimal number of clusters: 2 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>  asw</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 0.0000000 0.5729062 0.5363097 0.4923203 0.4629785 0.4425575 0.4417958
 [8] 0.4559577 0.4597152 0.4569222 0.4682526 0.4832332 0.4359832 0.4365490
[15] 0.4249453 0.4193369 0.4317815 0.4323037 0.4333972 0.4463586</code></pre>
</div>
</div>
<p>We rerun and save the results for a 7k cluster partition. If you inspect the resulting data frame, it provides an identifier for each cluster. Each individual area is attributed to a cluster. Next the question that we seek to answer is what sort of pattern do these clusters capture?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rerun pam for k=7</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>pam_optimal <span class="ot">&lt;-</span> <span class="fu">pam</span>(seq.om, <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualising" class="level3" data-number="4.3.4">
<h3 data-number="4.3.4" class="anchored" data-anchor-id="visualising"><span class="header-section-number">4.3.4</span> Visualising</h3>
<p>To understand the representative patterns captured in our data partition, we use visualisation. There is a battery of different visualisation tools to extract information and identify distinctive features of the identified trajectories. We start by using <em>individual sequence plots</em> by trajectory type. They provide a visual representation of how individual areas in each trajectory type moves between states. Recall that we are capturing representative trajectories; hence, there is still quite a bit of variability in terms of the patterns encapsulated in each representative trajectory. Back to the individual sequence plots, each line in these plots represents an area. Time is displayed horizontally and colours encode different states - in our case of population change. Numbers on the y-axis display the number of areas in each cluster. The figure immediate below relies on the base <code>plot</code> library, and by default, it is not very visually appealing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create individual sequence plots</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">seqplot</span>(seq.cl, </span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> pam_optimal<span class="sc">$</span>clustering,</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="st">"I"</span>,</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">border =</span> <span class="cn">NA</span>, </span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex.axis =</span> <span class="fl">1.5</span>, </span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex.lab =</span> <span class="fl">1.5</span>,</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">sortv =</span> seq.om)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We therefore switch to the R library <code>ggseqplot</code> which enables visualisation of sequence data based on <code>ggplot</code> functionalities. This package may provide more flexibility if we are more familiar with <code>ggplot</code>.</p>
<p>The figure below offers a clear representation of the systematic sequencing of states that each trajectory captures. It provides information on two key features of trajectories: sequencing and size. For example, trajectory 1 seems to capture a sequencing pattern of transitions from moderate population decline to stability and back to moderate decline. Trajectory 2 shows a pattern high population decline during the first few years and then consistent moderate decline. Trajectory 3 displays a predominant pattern of moderate population decline. Trajectory 4 represents patterns of areas experiencing decline with spells of high population decline. Trajectory 5 shows a pattern of decline in the first few years followed by moderate decline and decline again. Trajectory 6 shows a similar pattern with more prevalent spells of population decline across the entire period. Trajectory 7 displays a trend of temporary decline, with spells of population growth and stability. From these plots, you can also identify which trajectories tend to be more common. In our example, trajectory and 3 accounts the largest number of areas: 189.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create individual sequence plots based on ggplot</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggseqiplot</span>(seq.cl, </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> pam_optimal<span class="sc">$</span>clustering,</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">sortv =</span> seq.om,</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">facet_ncol =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">rev</span>(cols)) <span class="sc">+</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">rev</span>(cols)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.
Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can also get a better understanding of the resulting trajectories by analysing <em>state frequency plots</em>. They show the number of occurrences of a given state in individual years. These plots examine the data from a vertical perspective i.e.&nbsp;looking at individual years across areas, rather than at individual areas over time. State frequency plots reveal that predominant states in each year and changes in their prevalence. Focusing on trajectory 1, for example, we observe that moderate decline was the predominant state between 2000 and 2007 and stability became equally prevalent during 2008 and 2015.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create state frequency plots based on ggplot</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggseqdplot</span>(seq.cl, </span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> pam_optimal<span class="sc">$</span>clustering,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">facet_ncol =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cols) <span class="sc">+</span></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> cols) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can also examine time spent in individual states in each trajectory. <em>Time spent plots</em> report the average time spent in each state. The measure of time depends on the original data used in the analysis. We use years so the y-axis refers to the average number of years that a given status appears in a representative trajectory type. For example, a score over 5 for stable in trajectory 1 indicates that the average number of years that areas in that typology are classified in that category is over 5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create time spent plots based on ggplot</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggseqmtplot</span>(seq.cl, </span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> pam_optimal<span class="sc">$</span>clustering,</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">facet_ncol =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">rev</span>(cols)) <span class="sc">+</span></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">rev</span>(cols)) <span class="sc">+</span></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">labels=</span><span class="fu">c</span>(<span class="st">"high_growth"</span> <span class="ot">=</span> <span class="st">"HG"</span>, <span class="st">"growth"</span> <span class="ot">=</span> <span class="st">"G"</span>,</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>                              <span class="st">"moderate_growth"</span> <span class="ot">=</span> <span class="st">"MG"</span>, <span class="st">"stable"</span> <span class="ot">=</span> <span class="st">"S"</span>, <span class="st">"moderate_decline"</span> <span class="ot">=</span> <span class="st">"MD"</span>, <span class="st">"decline"</span> <span class="ot">=</span> <span class="st">"D"</span>, <span class="st">"high_decline"</span> <span class="ot">=</span> <span class="st">"HD"</span> ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally we analyse <em>entropy index plots</em>. Entropy is a measure of diversity. The greater the score, the greater the entropy or diversity of states. The plot below displays the entropy index computed for individual trajectories each year. Each line represents the entropy index for a trajectory in each year. The top yellow line in 2001 indicates that in 2001 areas following a trajectory 7 type were distributed across a larger number of states than any other trajectory, reflecting the fact that areas experiencing this trajectory moves between states of decline, stability and growth. In other word, it indicates that there was more diversity of states.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create entropy index plots based on ggplot</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggseqeplot</span>(seq.cl, </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">group =</span> pam_optimal<span class="sc">$</span>clustering) <span class="sc">+</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_viridis_d</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="sequence-analysis_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="questions" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="questions"><span class="header-section-number">4.4</span> Questions</h2>
<p>For the first assignment, we will continue to focus on London as our area of analysis. We will use population count estimates from the Office of National Statistics (ONS). The dataset provides information on area, population numbers and population density at national, regional and smaller sub-national area level, including Unitary Authority, Metropolitan County, Metropolitan District, County, Non-metropolitan District, London Borough, Council Area and Local Government District for the period from 2001 to 2020.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>pop_df <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"./data/sequence-analysis/population_uk/population-uk-2011_20.csv"</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>pop_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 420 × 44
   Code    Name  Geogr…¹ Area …² Estim…³ 2020 …⁴ Estim…⁵ 2019 …⁶ Estim…⁷ 2018 …⁸
   &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 K02000… UNIT… Country  2.43e5  6.71e7   276.   6.68e7   275.   6.64e7   274. 
 2 K03000… GREA… Country  2.29e5  6.52e7   285.   6.49e7   283.   6.46e7   282. 
 3 K04000… ENGL… Country  1.51e5  5.97e7   395.   5.94e7   394.   5.91e7   391. 
 4 E92000… ENGL… Country  1.30e5  5.66e7   434.   5.63e7   432.   5.60e7   430. 
 5 E12000… NORT… Region   8.58e3  2.68e6   312.   2.67e6   311.   2.66e6   310. 
 6 E06000… Coun… Unitar…  2.23e3  5.33e5   240.   5.30e5   238.   5.27e5   237. 
 7 E06000… Darl… Unitar…  1.97e2  1.07e5   544.   1.07e5   541.   1.07e5   540. 
 8 E06000… Hart… Unitar…  9.37e1  9.38e4  1001.   9.37e4   999.   9.32e4   995. 
 9 E06000… Midd… Unitar…  5.39e1  1.41e5  2622.   1.41e5  2616.   1.41e5  2608. 
10 E06000… Nort… Unitar…  5.02e3  3.24e5    64.5  3.22e5    64.2  3.20e5    63.8
# … with 410 more rows, 34 more variables:
#   `Estimated Population mid-2017` &lt;dbl&gt;, `2017 people per sq. km` &lt;dbl&gt;,
#   `Estimated Population mid-2016` &lt;dbl&gt;, `2016 people per sq. km` &lt;dbl&gt;,
#   `Estimated Population mid-2015` &lt;dbl&gt;, `2015 people per sq. km` &lt;dbl&gt;,
#   `Estimated Population mid-2014` &lt;dbl&gt;, `2014 people per sq. km` &lt;dbl&gt;,
#   `Estimated Population mid-2013` &lt;dbl&gt;, `2013 people per sq. km` &lt;dbl&gt;,
#   `Estimated Population mid-2012` &lt;dbl&gt;, `2012 people per sq. km` &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>For the assignment, you should only work with smaller sub-national areas. Filter out country and regional area. You should address the following questions:</p>
<ol type="1">
<li><p>Use sequence analysis to identify representative trajectories of population change and discuss the type of trajectories identified in London Boroughs.</p></li>
<li><p>Use individual sequence plot to identify distinctive features in the resulting trajectories.</p></li>
</ol>
<p>For the analysis, aim to focus on the area of London so you can link your narrative to the rest of analyses you will be conducting.</p>
<p>Ensure you justify the number of optimal clusters you will use in your analysis and provide a brief description of the trajectories identified. Describe how they are unique.</p>


<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography" style="display: none">
<div id="ref-abbott2000sequence" class="csl-entry" role="doc-biblioentry">
Abbott, Andrew, and Angela Tsay. 2000. <span>“Sequence Analysis and Optimal Matching Methods in Sociology: Review and Prospect.”</span> <em>Sociological Methods &amp; Research</em> 29 (1): 3–33.
</div>
<div id="ref-backman2020" class="csl-entry" role="doc-biblioentry">
Backman, Mikaela, Esteban Lopez, and Francisco Rowe. 2020. <span>“The Occupational Trajectories and Outcomes of Forced Migrants in Sweden. Entrepreneurship, Employment or Persistent Inactivity?”</span> <em>Small Business Economics</em> 56 (3): 963–83. <a href="https://doi.org/10.1007/s11187-019-00312-z">https://doi.org/10.1007/s11187-019-00312-z</a>.
</div>
<div id="ref-fielding1992" class="csl-entry" role="doc-biblioentry">
Fielding, A. J. 1992. <span>“Migration and Social Mobility: South East England as an Escalator Region.”</span> <em>Regional Studies</em> 26 (1): 1–15. <a href="https://doi.org/10.1080/00343409212331346741">https://doi.org/10.1080/00343409212331346741</a>.
</div>
<div id="ref-gabadinho2011" class="csl-entry" role="doc-biblioentry">
Gabadinho, Alexis, Gilbert Ritschard, Nicolas S. Müller, and Matthias Studer. 2011. <span>“Analyzing and Visualizing State Sequences in<span><em>R</em></span>with<span><strong>TraMineR</strong></span>.”</span> <em>Journal of Statistical Software</em> 40 (4). <a href="https://doi.org/10.18637/jss.v040.i04">https://doi.org/10.18637/jss.v040.i04</a>.
</div>
<div id="ref-gabadinho2009mining" class="csl-entry" role="doc-biblioentry">
Gabadinho, Alexis, Gilbert Ritschard, Matthias Studer, and Nicolas S Müller. 2009. <span>“Mining Sequence Data in r with the TraMineR Package: A User’s Guide.”</span> <em>Geneva: Department of Econometrics and Laboratory of Demography, University of Geneva</em>.
</div>
<div id="ref-gonzález-leonardo2023" class="csl-entry" role="doc-biblioentry">
González-Leonardo, Miguel, Niall Newsham, and Francisco Rowe. 2023. <span>“Understanding Population Decline Trajectories in Spain Using Sequence Analysis.”</span> <em>Geographical Analysis</em>, January. <a href="https://doi.org/10.1111/gean.12357">https://doi.org/10.1111/gean.12357</a>.
</div>
<div id="ref-kaufman2009finding" class="csl-entry" role="doc-biblioentry">
Kaufman, Leonard, and Peter J Rousseeuw. 2009. <em>Finding Groups in Data: An Introduction to Cluster Analysis</em>. John Wiley &amp; Sons.
</div>
<div id="ref-newsham2022" class="csl-entry" role="doc-biblioentry">
Newsham, Niall, and Francisco Rowe. 2022a. <span>“Understanding the Trajectories of Population Decline Across Rural and Urban Europe: A Sequence Analysis.”</span> <a href="https://doi.org/10.48550/ARXIV.2203.09798">https://doi.org/10.48550/ARXIV.2203.09798</a>.
</div>
<div id="ref-newsham2022a" class="csl-entry" role="doc-biblioentry">
———. 2022b. <span>“Understanding Trajectories of Population Decline Across Rural and Urban Europe: A Sequence Analysis.”</span> <em>Population, Space and Place</em>, December. <a href="https://doi.org/10.1002/psp.2630">https://doi.org/10.1002/psp.2630</a>.
</div>
<div id="ref-patias2021" class="csl-entry" role="doc-biblioentry">
Patias, Nikos, Francisco Rowe, and Dani Arribas-Bel. 2021. <span>“Trajectories of Neighbourhood Inequality in Britain: Unpacking Inter<span>-</span>Regional Socioeconomic Imbalances, 1971<span>-</span>2011.”</span> <em>The Geographical Journal</em> 188 (2): 150–65. <a href="https://doi.org/10.1111/geoj.12420">https://doi.org/10.1111/geoj.12420</a>.
</div>
<div id="ref-patias2021a" class="csl-entry" role="doc-biblioentry">
Patias, Nikos, Francisco Rowe, Stefano Cavazzi, and Dani Arribas-Bel. 2021. <span>“Sustainable Urban Development Indicators in Great Britain from 2001 to 2016.”</span> <em>Landscape and Urban Planning</em> 214 (October): 104148. <a href="https://doi.org/10.1016/j.landurbplan.2021.104148">https://doi.org/10.1016/j.landurbplan.2021.104148</a>.
</div>
<div id="ref-rowe2022e" class="csl-entry" role="doc-biblioentry">
Rowe, Francisco, and Dani Arribas-Bel. 2022. <span>“Spatial Modelling for Data Scientists.”</span> <em>Open Science Framework</em>, August. <a href="https://doi.org/10.17605/OSF.IO/8F6XR">https://doi.org/10.17605/OSF.IO/8F6XR</a>.
</div>
<div id="ref-rowe2016" class="csl-entry" role="doc-biblioentry">
Rowe, Francisco, Jonathan Corcoran, and Martin Bell. 2016. <span>“The Returns to Migration and Human Capital Accumulation Pathways: Non-Metropolitan Youth in the School-to-Work Transition.”</span> <em>The Annals of Regional Science</em> 59 (3): 819–45. <a href="https://doi.org/10.1007/s00168-016-0771-8">https://doi.org/10.1007/s00168-016-0771-8</a>.
</div>
<div id="ref-salmela-aro2011" class="csl-entry" role="doc-biblioentry">
Salmela-Aro, Katariina, Noona Kiuru, Jari-Erik Nurmi, and Mervi Eerola. 2011. <span>“Mapping Pathways to Adulthood Among Finnish University Students: Sequences, Patterns, Variations in Family- and Work-Related Roles.”</span> <em>Advances in Life Course Research</em> 16 (1): 25–41. <a href="https://doi.org/10.1016/j.alcr.2011.01.003">https://doi.org/10.1016/j.alcr.2011.01.003</a>.
</div>
<div id="ref-tatem2017" class="csl-entry" role="doc-biblioentry">
Tatem, Andrew J. 2017. <span>“WorldPop, Open Data for Spatial Demography.”</span> <em>Scientific Data</em> 4 (1). <a href="https://doi.org/10.1038/sdata.2017.4">https://doi.org/10.1038/sdata.2017.4</a>.
</div>
<div id="ref-turok2007" class="csl-entry" role="doc-biblioentry">
Turok, Ivan, and Vlad Mykhnenko. 2007. <span>“The Trajectories of European Cities, 1960<span></span>2005.”</span> <em>Cities</em> 24 (3): 165–82. <a href="https://doi.org/10.1016/j.cities.2007.01.007">https://doi.org/10.1016/j.cities.2007.01.007</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./geodemographics.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Geodemographics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./network.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Network Analysis</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Population Science was written by Francisco Rowe, Carmen Cabrera-Arnau, Elisabetta Pietrostefani.</div>   
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>



</body></html>